tools/shared.py

```diff
@@ -2756,6 +2757,27 @@ def wasm2js(js_file, wasm_file, opt_level, minify_whitespace, use_closure_compil
       f.write(all_js)
     return js_file
 
+  @staticmethod
+  def emit_debug_on_side(wasm_file):
+    # extract the DWARF info from the main file, and leave the wasm with
+    # debug into as a file on the side
+    # TODO: emit only debug sections in the side file, and not the entire
+    #       wasm as well
+    wasm_file_with_dwarf = wasm_file + '.debug.wasm'
+    shutil.move(wasm_file, wasm_file_with_dwarf)
+    run_process([LLVM_OBJCOPY, '--remove-section=.debug*', wasm_file_with_dwarf, wasm_file])
+
+    # embed a section in the main wasm to point to the file with external DWARF,
+    # see https://yurydelendik.github.io/webassembly-dwarf/#external-DWARF
+    section_name = b'\x13external_debug_info' # section name, including prefixed size
+    contents = asbytes(wasm_file_with_dwarf)
+    section_size = len(section_name) + len(contents)
+    with open(wasm_file, 'ab') as f:
+      f.write(b'\0') # user section is code 0
```

Comment:
This bit isn't necessary. The section header is created by objcopy. We only need the payload here.

Reply:
(the alternative of course would be to create the header here too and then just directly append to the file rather than using objcopy at all)

Reply:
This doesn't use `objcopy` to add the new section - it just appends the bytes itself manually, so we emit all the parts. (We do similar things for a few other custom sections, like the dylink section, see `make_shared_library()`.)

Reply:
oh oops, I didn't read it closely enough :D